File: packages/zynth-ui/ios/ZynthUIModule.swift
import UIKit
import ZynthKit

@objc(ZynthUIModule)
final class ZynthUIModule: NSObject, ZynthModule, ZynthSyncModule {

  let name = "ZynthUI"

  var exportedMethods: [String] {
    return ["generateQRCode"]
  }

  func call(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "generateQRCode":
      let payload = try generateQRCode(args: args)
      return ["result": payload]
    default:
      throw ZynthModuleError.methodNotExported(module: name, method: method)
    }
  }

  func callSync(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "generateQRCode":
      return try generateQRCode(args: args)
    default:
      throw ZynthModuleError.methodNotExported(module: name, method: method)
    }
  }

  private func generateQRCode(args: ZynthArgs) throws -> [String: Any] {
    let size = max(Int(args.number("size", default: 192)), 1)
    let matrix = try parseMatrix(args: args)
    if matrix.isEmpty {
      throw NSError(
        domain: "ZynthUI",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "QR matrix cannot be empty"]
      )
    }

    let darkColorRaw = args.optionalString("color") ?? "#000000"
    let lightColorRaw = args.optionalString("backgroundColor") ?? "#FFFFFF"
    let darkColor = parseColor(darkColorRaw) ?? .black
    let lightColor = parseColor(lightColorRaw) ?? .white

    guard let image = rasterizeMatrix(
      matrix: matrix,
      size: size,
      darkColor: darkColor,
      lightColor: lightColor
    ) else {
      throw NSError(
        domain: "ZynthUI",
        code: 2,
        userInfo: [NSLocalizedDescriptionKey: "Unable to render QR image"]
      )
    }

    guard let pngData = image.pngData() else {
      throw NSError(
        domain: "ZynthUI",
        code: 3,
        userInfo: [NSLocalizedDescriptionKey: "Unable to encode QR image"]
      )
    }

    return [
      "imageData": pngData.base64EncodedString(),
      "mimeType": "image/png",
      "width": size,
      "height": size
    ]
  }

  private func parseMatrix(args: ZynthArgs) throws -> [[Bool]] {
    let rawRows = try args.array("matrix")
    if rawRows.isEmpty {
      return []
    }

    var rows: [[Bool]] = []
    var expectedColumns = -1

    for rawRow in rawRows {
      guard let rowArray = rawRow as? [Any] else {
        throw NSError(
          domain: "ZynthUI",
          code: 4,
          userInfo: [NSLocalizedDescriptionKey: "QR matrix rows must be arrays"]
        )
      }

      if expectedColumns < 0 {
        expectedColumns = rowArray.count
      }
      if rowArray.count != expectedColumns {
        throw NSError(
          domain: "ZynthUI",
          code: 5,
          userInfo: [NSLocalizedDescriptionKey: "QR matrix must be square"]
        )
      }

      var rowValues: [Bool] = []
      rowValues.reserveCapacity(rowArray.count)
      for cell in rowArray {
        if let boolValue = cell as? Bool {
          rowValues.append(boolValue)
        } else if let number = cell as? NSNumber {
          rowValues.append(number.intValue != 0)
        } else {
          rowValues.append(false)
        }
      }
      rows.append(rowValues)
    }

    if rows.count != expectedColumns {
      throw NSError(
        domain: "ZynthUI",
        code: 6,
        userInfo: [NSLocalizedDescriptionKey: "QR matrix must be square"]
      )
    }

    return rows
  }

  private func rasterizeMatrix(
    matrix: [[Bool]],
    size: Int,
    darkColor: UIColor,
    lightColor: UIColor
  ) -> UIImage? {
    let moduleCount = matrix.count
    if moduleCount == 0 {
      return nil
    }

    let modulePixel = max(size / moduleCount, 1)
    let drawSize = modulePixel * moduleCount
    let drawOffset = max((size - drawSize) / 2, 0)
    let rect = CGRect(x: 0, y: 0, width: size, height: size)

    UIGraphicsBeginImageContextWithOptions(rect.size, true, 1)
    guard let context = UIGraphicsGetCurrentContext() else {
      UIGraphicsEndImageContext()
      return nil
    }

    context.setFillColor(lightColor.cgColor)
    context.fill(rect)
    context.setFillColor(darkColor.cgColor)

    for row in 0..<moduleCount {
      let values = matrix[row]
      var col = 0
      while col < values.count {
        if !values[col] {
          col += 1
          continue
        }
        let start = col
        while col < values.count && values[col] {
          col += 1
        }
        let width = (col - start) * modulePixel
        let x = drawOffset + start * modulePixel
        let y = drawOffset + row * modulePixel
        context.fill(CGRect(x: x, y: y, width: width, height: modulePixel))
      }
    }

    let image = UIGraphicsGetImageFromCurrentImageContext()
    UIGraphicsEndImageContext()
    return image
  }

  private func parseColor(_ raw: String) -> UIColor? {
    let value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    if value.isEmpty {
      return nil
    }

    if value.hasPrefix("#") {
      let hex = String(value.dropFirst())
      if hex.count == 6 {
        var parsed: UInt64 = 0
        guard Scanner(string: hex).scanHexInt64(&parsed) else {
          return nil
        }
        let red = CGFloat((parsed >> 16) & 0xFF) / 255.0
        let green = CGFloat((parsed >> 8) & 0xFF) / 255.0
        let blue = CGFloat(parsed & 0xFF) / 255.0
        return UIColor(red: red, green: green, blue: blue, alpha: 1.0)
      }

      if hex.count == 8 {
        var parsed: UInt64 = 0
        guard Scanner(string: hex).scanHexInt64(&parsed) else {
          return nil
        }
        let alpha = CGFloat((parsed >> 24) & 0xFF) / 255.0
        let red = CGFloat((parsed >> 16) & 0xFF) / 255.0
        let green = CGFloat((parsed >> 8) & 0xFF) / 255.0
        let blue = CGFloat(parsed & 0xFF) / 255.0
        return UIColor(red: red, green: green, blue: blue, alpha: alpha)
      }
    }

    return nil
  }
}

File: packages/zynth-ui/android/ZynthUi/src/main/java/dev/zynth/ui/ZynthUIModule.kt
package dev.zynth.ui

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.Base64
import android.util.Log
import com.zynth.kit.runtime.ZynthArgs
import com.zynth.kit.runtime.ZynthModule
import com.zynth.kit.runtime.ZynthSyncModule
import org.json.JSONObject
import java.io.ByteArrayOutputStream

/**
 * Native UI utility module.
 *
 * Add generic UI-level native helpers here as they are introduced.
 */
class ZynthUIModule : ZynthModule, ZynthSyncModule {
    companion object {
        private const val TAG = "ZynthUI.QR"
    }

    override val name: String = "ZynthUI"

    override val exportedMethods: List<String> = listOf("generateQRCode")

    override fun call(method: String, args: ZynthArgs): JSONObject {
        return when (method) {
            "generateQRCode" -> resultResponse(generateQRCodeObject(args))
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
    }

    override fun callSync(method: String, args: ZynthArgs): Any {
        return when (method) {
            "generateQRCode" -> {
                val result = generateQRCodeObject(args)
                Log.d(
                    TAG,
                    "callSync generateQRCode -> keys=${result.keys().asSequence().toList()} " +
                        "hasImageData=${result.optString("imageData").isNotBlank()} " +
                        "mime=${result.optString("mimeType")} " +
                        "width=${result.optInt("width")} height=${result.optInt("height")}"
                )
                result
            }
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
    }

    private fun generateQRCodeObject(args: ZynthArgs): JSONObject {
        val size = args.getInt("size", 192).coerceAtLeast(1)
        val matrix = parseMatrix(args)
        val moduleCount = matrix.size
        if (moduleCount == 0) {
            throw IllegalArgumentException("QR matrix cannot be empty")
        }

        val darkColor = parseColor(args.getString("color", "#000000"), Color.BLACK)
        val lightColor = parseColor(args.getString("backgroundColor", "#FFFFFF"), Color.WHITE)
        Log.d(
            TAG,
            "generateQRCode size=$size moduleCount=$moduleCount " +
                "darkColor=${Integer.toHexString(darkColor)} lightColor=${Integer.toHexString(lightColor)}"
        )

        val bitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        canvas.drawColor(lightColor)

        val modulePixelSize = (size / moduleCount).coerceAtLeast(1)
        val drawSize = modulePixelSize * moduleCount
        val drawOffset = ((size - drawSize) / 2).coerceAtLeast(0)

        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = darkColor
            style = Paint.Style.FILL
        }

        for (row in 0 until moduleCount) {
            val values = matrix[row]
            var col = 0
            while (col < values.size) {
                if (!values[col]) {
                    col += 1
                    continue
                }
                val start = col
                while (col < values.size && values[col]) {
                    col += 1
                }
                val left = drawOffset + start * modulePixelSize
                val top = drawOffset + row * modulePixelSize
                val width = (col - start) * modulePixelSize
                canvas.drawRect(
                    left.toFloat(),
                    top.toFloat(),
                    (left + width).toFloat(),
                    (top + modulePixelSize).toFloat(),
                    paint,
                )
            }
        }

        val output = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, output)
        bitmap.recycle()

        val bytes = output.toByteArray()
        val encoded = Base64.encodeToString(bytes, Base64.NO_WRAP)
        Log.d(TAG, "generateQRCode encoded pngBytes=${bytes.size} base64Len=${encoded.length}")

        return JSONObject()
            .put("imageData", encoded)
            .put("mimeType", "image/png")
            .put("width", size)
            .put("height", size)
    }

    private fun parseMatrix(args: ZynthArgs): List<List<Boolean>> {
        val rawRows = args.getList("matrix")
        if (rawRows.isEmpty()) {
            throw IllegalArgumentException("QR matrix cannot be empty")
        }

        val rows = mutableListOf<List<Boolean>>()
        var expectedColumns = -1

        for (rawRow in rawRows) {
            if (rawRow !is List<*>) {
                throw IllegalArgumentException("QR matrix rows must be arrays")
            }
            if (expectedColumns == -1) {
                expectedColumns = rawRow.size
            }
            if (rawRow.size != expectedColumns) {
                throw IllegalArgumentException("QR matrix must be square")
            }

            val parsedRow = mutableListOf<Boolean>()
            for (cell in rawRow) {
                val isDark = when (cell) {
                    is Boolean -> cell
                    is Number -> cell.toInt() != 0
                    else -> false
                }
                parsedRow.add(isDark)
            }
            rows.add(parsedRow)
        }

        if (rows.size != expectedColumns) {
            throw IllegalArgumentException("QR matrix must be square")
        }

        Log.d(TAG, "parseMatrix rows=${rows.size} cols=$expectedColumns")
        return rows
    }

    private fun parseColor(value: String, fallback: Int): Int {
        return try {
            Color.parseColor(value)
        } catch (_: IllegalArgumentException) {
            fallback
        }
    }

    private fun resultResponse(result: Any?): JSONObject {
        return JSONObject().put("result", result)
    }
}
